File: Έγγραφα SEM4105/Λύσεις ασκήσεων/03. Τρίτο εργαστήριο/2a/iteration1.f
C  Fixed point method to locate a root of 
C   
C   f(x)=x^2-6x+5
C
     PROGRAM FIXPOI

      DOUBLE PRECISION X

C     declare that F,G are functions
      EXTERNAL F,G

C     function names type declaration
      DOUBLE PRECISION F,G

C     function value
      DOUBLE PRECISION FNVAL

C     a small constant number
      DOUBLE PRECISION TOLER
      PARAMETER (TOLER = 1D-8)

C     a very large number
      DOUBLE PRECISION VFAR
      PARAMETER (VFAR = 1D5)

C     initial guess for x
      X = 1.3D0

 100  FNVAL = F(X)

      PRINT *, "The current approximation is ", X 
      PRINT *, "The current value of F(x) is ", FNVAL
      PRINT *

      IF (ABS(X) .GT. VFAR) THEN
         PRINT *, "X is very large; probably g(x) is not appropriate"
         GOTO 200
      ENDIF

      IF (ABS(FNVAL) .LT. TOLER) GOTO 200

      X = G(X)
      GOTO 100

 200  END


      FUNCTION F(X)
      DOUBLE PRECISION X,F

      F = X**2 - 6.0D0 * X + 5.0D0
      END


C     if f(x) above then 
C     f(x) = 0 => x = g(x) 
C     with g as follows:

      FUNCTION G(X)
      DOUBLE PRECISION X,G

      G = (X**2 + 5.0D0) / 6.0D0
      END

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/03. Τρίτο εργαστήριο/2a/iteration1.c
#include <stdio.h>
#include <math.h>

double f(double x)
{
  return x*x - 6.0 * x + 5.0;
}



/*  
    if f(x) above then 
    f(x) = 0 => x = g(x) 
    with g as follows:
*/
double g(double x)
{
  return (x*x + 5.0) / 6.0;
}




/*
  Fixed point method to locate a root of 
   
  f(x)=x^2-6x+5
*/


int
main(void)
{
  double const toler = 1E-8;   //  a small constant number
  double const vfar = 1e5;  // a very large number

  double x = 1.3;  // initial guess for x
  while (1) {
    double fnval = f(x); // function value 
	
    if (fabs(x) > vfar) {
      printf("X is very large; probably g(x) is not appropriate\n");
      break;
    }

    printf("The current approximation is %.12g\n", x);
    printf("The current value of F(x) is %g\n\n", fnval);

    if (fabs(fnval) < toler) {
      break;
    }

    x = g(x);
  }

  return 0;
}
